### PR TITLE
(impl) oAuth and Google

### PR DESCRIPTION
**Pull request by Identityofsine**
oAuth2 integration with future support for facebook and microsoft

## Copilot
This pull request introduces significant changes to the codebase, focusing on adding support for OAuth2 authentication (specifically Google OAuth2), improving the authentication flow, and enhancing code organization. It includes updates to configuration files, new service and controller classes, and modifications to existing authentication-related components.

### OAuth2 Authentication Support:

* **Google OAuth2 Integration**:
  - Added `GoogleOAuthService` to handle Google OAuth2 authentication, including methods for redirect URL generation and ID token exchange. (`src/main/java/com/prolinkli/core/app/components/oauth2/service/GoogleOAuthService.java`)
  - Introduced `OAuth2Controller` to manage OAuth2 login and callback endpoints. (`src/main/java/com/prolinkli/core/app/components/oauth2/controller/OAuth2Controller.java`)
  - Updated `Constants` to include Google OAuth2-related keys and authentication methods. (`src/main/java/com/prolinkli/core/app/Constants.java`) [[1]](diffhunk://#diff-e01baac7fe840f8fc37049ed56395c415e1f0f765d2633e3dc714e7545b69a86R22-R37) [[2]](diffhunk://#diff-e01baac7fe840f8fc37049ed56395c415e1f0f765d2633e3dc714e7545b69a86R61-R68)

### Configuration Updates:

* **Dependency Management**:
  - Added `google-api-client` dependency for Google OAuth2 integration. (`pom.xml`)
  - Included `spring-webflux` and `spring-boot-starter-thymeleaf` dependencies for enhanced web capabilities. (`pom.xml`) [[1]](diffhunk://#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R40-R43) [[2]](diffhunk://#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R120-R123)
* **Environment Variables**:
  - Removed `.env` file and updated `env.template` with Google OAuth2 configuration placeholders. (`env.template`) [[1]](diffhunk://#diff-e9cbb0224c4a3d23a6019ba557e0cd568c1ad5e1582ff1e335fb7d99b7a1055dL1-L28) [[2]](diffhunk://#diff-bb1592d4d5db99dab806ea26e6023803146aaea8e5c2b35e800822d6daba512fL20-R23)

### Authentication Enhancements:

* **User Authentication Updates**:
  - Added support for Google OAuth2 login in `UserController`. (`src/main/java/com/prolinkli/core/app/components/user/controller/UserController.java`)
  - Enhanced `UserAuthService` to handle Google OAuth2 credentials and updated authentication logic. (`src/main/java/com/prolinkli/core/app/components/user/service/UserAuthService.java`) [[1]](diffhunk://#diff-3e0a83a8d4f0ab2b2479dd7dcf20a84828ccca4a30b9cf665d61c5315d51764cR48-R55) [[2]](diffhunk://#diff-3e0a83a8d4f0ab2b2479dd7dcf20a84828ccca4a30b9cf665d61c5315d51764cL66-R86)
* **Logout Endpoint**:
  - Introduced a logout endpoint in `UserController` to clear authentication cookies. (`src/main/java/com/prolinkli/core/app/components/user/controller/UserController.java`)

### Code Organization:

* **Abstract Service Class**:
  - Created `AbstractOAuthService` to serve as a base class for OAuth2 providers. (`src/main/java/com/prolinkli/core/app/components/oauth2/service/AbstractOAuthService.java`)
* **UserAuthenticationForm Enhancements**:
  - Modified `UserAuthenticationForm` to encapsulate parameters more securely and added getter/setter restrictions. (`src/main/java/com/prolinkli/core/app/components/user/model/UserAuthenticationForm.java`) [[1]](diffhunk://#diff-7ad77f8b2430acc008ec009282c6884b4a0ac94dee1fc178ac4c6ca81c80ca30R10) [[2]](diffhunk://#diff-7ad77f8b2430acc008ec009282c6884b4a0ac94dee1fc178ac4c6ca81c80ca30L19-R27)

These changes collectively improve the application's authentication capabilities, particularly by enabling OAuth2 support, while enhancing maintainability and scalability.

### CODE REVIEW

File: src/main/java/com/prolinkli/core/app/components/oauth2/service/GoogleOAuthService.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import com.prolinkli.core.app.Constants.OAuth2Providers;
+import com.prolinkli.core.app.Constants.OAuth2ProvidersLks;
+import org.springframework.stereotype.Service;
+
+import jakarta.servlet.http.HttpServletRequest;
+
+@Service
+public class GoogleOAuthService extends AbstractOAuthService {
+
+  private static final String REDIRECT_URL = "http://localhost:8080/v1/api/auth/oauth2/google/callback";
```
[the remark this answers is not in the document]
Comment (Identityofsine) [reply src/main/java/com/prolinkli/core/app/components/oauth2/service/GoogleOAuthService.java:51]: @ColeBaxendale moving this to a different ticket so just ignore this one; I want to be able to setup secrets in a more reliable way so we aren't sharing one .env file

File: src/main/java/com/prolinkli/core/app/components/user/controller/UserController.java [hunk 2]
```diff
@@ -98,4 +172,20 @@ public User getCurrentUser(@CurrentUser User user) {
     return AuthorizedUser.strip(user);
   }
 
+  /**
+   * Logout endpoint that clears authentication cookies.
+   * 
+   * @param response HttpServletResponse to clear cookies.
+   * @return Success message.
+   */
+  @RequestMapping("/logout")
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/core/app/components/user/controller/UserController.java:181]: Switch to @PostMapping or @DeleteMapping to prevent accidental logout via GET requests.

File: src/main/java/com/prolinkli/core/app/components/user/controller/UserController.java [hunk 3]
```diff
@@ -98,4 +172,20 @@ public User getCurrentUser(@CurrentUser User user) {
...
+  public ResponseObject<Void> logout(HttpServletResponse response) {
+    // Clear authentication cookies
+    cookieSaveService.saveCookies(
+        JwtCookieUtil.createClearAuthCookies(),
+        response);
+
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/core/app/components/user/controller/UserController.java:187]: 🔒 SECURITY: Add token invalidation to logout

Current logout only clears cookies, leaving stolen tokens valid until expiration. Add server-side token invalidation to immediately block old tokens:

```java
// Add to logout method:
if (user != null) {
    jwtInvalidationService.invalidateTokens(user.getAuthToken());
}
```

This ensures logout actually stops unauthorized access by immediately invalidating tokens in the database.

File: src/main/java/com/prolinkli/core/app/components/user/service/UserAuthService.java [hunk 4]
```diff
@@ -63,26 +65,24 @@ public AuthorizedUser login(UserAuthenticationForm userAuthForm) {
...
     if (user == null) {
-      throw new IllegalStateException("User not authenticated");
+      throw new AuthorizationDeniedException("User not authenticated");
     }
 
     return jwtSaveService.regenerateTokens(user, response);
   }
 
-  private Map<String, Object> getCredentials(UserAuthenticationForm userAuthForm) {
+  private void getCredentials(UserAuthenticationForm userAuthForm) {
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/core/app/components/user/service/UserAuthService.java:74]: The `getCredentials` method could be improved:

1. **Missing validation**: No handling for unknown authentication methods - should throw an exception for unsupported auth types
2. **Parameter mapping**: Could be more robust with null checks and validation of required fields before mapping
3. **Architecture**: Consider using a strategy pattern instead of if/else chains - would make adding new auth methods much cleaner

Current approach assumes valid auth method and required fields, which could cause runtime issues.

File: src/main/java/com/prolinkli/core/app/components/user/service/UserAuthService.java [hunk 5]
```diff
@@ -44,13 +45,14 @@ public AuthorizedUser login(UserAuthenticationForm userAuthForm) {
     }
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/core/app/components/user/service/UserAuthService.java:46]: The `login` method has several issues:

1. **Returns null on auth failure** - should throw `AuthenticationFailedException` instead of returning null
2. **Generic exception handling** -  Maybe? `RuntimeException` is too broad, should use a specific auth exception
3. **Missing user validation** - no check if `authForm.getUserFromCredentials()` returns null before JWT creation
4. **Inconsistent error handling** - some paths throw exceptions, others return null

Suggested improvements:
- Replace `return null` with proper exception
- Add null check for user after authentication
- Use specific authentication exceptions instead of generic RuntimeException (maybe)
- Consider logging authentication failures for security monitoring

File: src/main/java/com/prolinkli/core/app/components/user/service/UserCreateService.java [hunk 6]
```diff
@@ -54,61 +62,26 @@ public AuthorizedUser createUser(UserAuthenticationForm user) {
...
-      rollback(user);
-      LOGGER.error("Failed to insert credentials for user: {}:{}", user.getId(), user.getUsername());
-      throw new RuntimeException("Failed to insert credentials for user: " + user.getUsername(), e);
-    }
 
-    // Return the authorized user.
-    return userAuthService.login(user);
-
-  }
+      LOGGER.error("Failed to create user: {}:{}", user.getId(), user.getUsername(), e);
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/core/app/components/user/service/UserCreateService.java:78]: UserCreateService has potential null pointer issues:

1. `user.getId()` could be null when logging - `authProvider.createUser()` might not set the user ID
2.  Rollback logic checks `user.getId() != null` but the ID might never be set by the auth provider

The `createUser` method on auth providers may not populate the user ID, leading to null values in logging and rollback logic. Consider:
- Adding null checks before logging user.getId()
- Using @Transactional for proper rollback instead of manual rollback
- Ensuring auth providers properly set user IDs after creation

File: src/main/java/com/prolinkli/core/app/components/user/service/UserCreateService.java [hunk 7]
```diff
@@ -54,61 +62,26 @@ public AuthorizedUser createUser(UserAuthenticationForm user) {
...
-      throw new IllegalArgumentException("User already exists with username: " + user.getUsername());
-    }
-
-    validateUserName(user.getUsername());
 
-    Map<String, Object> credentials = Map.of(
-        AuthenticationKeys.PASSWORD.USERNAME, user.getUsername(),
-        AuthenticationKeys.PASSWORD.PASSWORD, user.getSpecialToken());
+      // Attempt to create an OAuth user if the method is OAuth2.
+      authProvider.createUser(user, dao);
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/core/app/components/user/service/UserCreateService.java:68]: Issue: If createUser fails partway through, there's no guarantee of proper cleanup.

**Problem**: AuthProvider.createUser() methods perform multiple database operations (user insert + credentials insert) without transaction boundaries. If the second operation fails, the first operation isn't rolled back.

**Solution**: Add @Transactional to createUser methods or wrap operations in explicit transaction management.

File: src/main/java/com/prolinkli/framework/auth/providers/GoogleOAuth2Provider.java [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+      // Check if user already exists
+      User foundUser = null;
+      try {
+        foundUser = userGetService.getUserByUsername(systemUsername);
+      } catch (IllegalArgumentException e) {
+        // User does not exist, proceed with creation.
+      }
+
+      if (foundUser != null) {
+        throw new IllegalArgumentException("User already exists with username: " + systemUsername);
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/framework/auth/providers/GoogleOAuth2Provider.java:213]: Should use ResourceAlreadyExists instead of IllegalArgumentException

File: src/main/java/com/prolinkli/framework/auth/providers/GoogleOAuth2Provider.java [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+
+      AuthValidationUtil.validateUserName(systemUsername);
+
+      // Insert the user into the database
+      UserDb userDb = new UserDb();
+      userDb.setUsername(systemUsername);
+      userDb.setAuthenticationMethod(user.getAuthenticationMethodLk().toUpperCase());
+
+      LOGGER.debug("Inserting OAuth user into database: {}", systemUsername);
+      dao.insert(userDb);
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/framework/auth/providers/GoogleOAuth2Provider.java:224]: Multiple database operations without transaction boundaries 

dao.insert(userDb);  // Operation 1
insertCredentialsForUser(user, Map.of(...));  // Operation 2

File: src/main/java/com/prolinkli/framework/auth/providers/GoogleOAuth2Provider.java [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+
+      // For OAuth users, we don't need to store credentials in the traditional way
+      // The OAuth provider handles authentication
+      LOGGER.info("Successfully created OAuth user: {}:{} with email: {}", user.getId(), systemUsername, email);
+
+      return;
+
+    } catch (Exception e) {
+      // Rollback the user creation if anything fails
+      throw new RuntimeException("Failed to create OAuth user: " + e.getMessage(), e);
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/framework/auth/providers/GoogleOAuth2Provider.java:246]: Generic exception catch that doesn't add value

File: src/main/java/com/prolinkli/framework/auth/providers/GoogleOAuth2Provider.java [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+      // Extract user information from Google ID token
+      String idToken = user.getSpecialToken();
+      var googleUserInfo = getGoogleUserInfo(idToken);
+
+      if (googleUserInfo == null) {
+        throw new IllegalArgumentException("Failed to extract user information from Google ID token");
+      }
+
+      String email = googleUserInfo.getEmail();
+      String googleUserId = googleUserInfo.getSubject();
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/framework/auth/providers/GoogleOAuth2Provider.java:194]: googleUserId could be null but no null check before use

File: src/main/java/com/prolinkli/framework/auth/providers/InternalAuthProvider.java [hunk 12]
```diff
@@ -35,17 +46,56 @@ public Boolean authenticate(Map<String, Object> credentials) {
...
+        AuthenticationKeys.PASSWORD.USERNAME, user.getUsername(),
+        AuthenticationKeys.PASSWORD.PASSWORD, user.getSpecialToken());
+
+    // Insert the user into the database.
+    UserDb userDb = new UserDb();
+    userDb.setUsername(user.getUsername());
+    userDb.setAuthenticationMethod(user.getAuthenticationMethodLk().toUpperCase());
+
+    LOGGER.debug("Inserting password user into database: {}", user.getUsername());
+    dao.insert(userDb);
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/framework/auth/providers/InternalAuthProvider.java:85]: InternalAuthProvider: Missing transaction management in createUser method

Perform multiple DB operations (user insert + credentials insert) without transaction boundaries. If credentials insert fails, user record remains orphaned.

Add @Transactional to createUser method.

File: src/main/java/com/prolinkli/framework/auth/providers/InternalAuthProvider.java [hunk 13]
```diff
@@ -35,17 +46,56 @@ public Boolean authenticate(Map<String, Object> credentials) {
...
+
+    // Verify that the insert was successful and ID was generated
+    if (userDb.getId() == null) {
+      throw new RuntimeException("Database insert failed - no ID was generated for user: " + user.getUsername());
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/framework/auth/providers/InternalAuthProvider.java:89]: InternalAuthProvider: Generic RuntimeException

Should use specific exception like DatabaseOperationException instead of generic RuntimeException for database insert failures.

File: src/main/java/com/prolinkli/framework/auth/providers/InternalAuthProvider.java [hunk 14]
```diff
@@ -85,7 +135,22 @@ public void insertCredentialsForUser(User user, Map<String, Object> credentials)
...
+  public User getUserFromCredentials(UserAuthenticationForm userAuthForm) {
+
+    var credentials = userAuthForm.getParameters();
+    this.validateCredentials(credentials);
+
+    String username = credentials.get(AuthenticationKeys.PASSWORD.USERNAME).toString();
+
+    User user = userGetService.getUserByUsername(username);
+    if (user == null) {
+      throw new IllegalArgumentException("User not found for the provided username");
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/framework/auth/providers/InternalAuthProvider.java:150]: Uses IllegalArgumentException instead of ResourceNotFoundException

File: src/main/java/com/prolinkli/framework/auth/util/OAuthUsernameUtil.java [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    String normalizedEmail = email.split("@")[0]; // Use the part before '@' for username
+    normalizedEmail = normalizedEmail.toLowerCase().replaceAll("[^a-z0-9]", "");
+
+    String truncatedOauthId = oauthId.replaceAll("[^a-z0-9]", "").toLowerCase().substring(0,
+        Math.min(oauthId.length(), 10));
+
+    StringBuilder emailBuilder = new StringBuilder(normalizedEmail);
+    StringBuilder oauthIdBuilder = new StringBuilder(truncatedOauthId);
+
+    int lastTruncated = 0b0000;
```
Comment (ColeBaxendale) [inline src/main/java/com/prolinkli/framework/auth/util/OAuthUsernameUtil.java:74]: OAuthUsernameUtil: Overly complex truncation logic on lines 73-83

The alternating truncation with bitwise operations is unnecessarily complex. Consider simpler proportional truncation or prioritize email over OAuth ID.

**Review by ColeBaxendale — CHANGES_REQUESTED** [review]
Thanks for the PR! Here are my review comments:

## General Feedback
- [ ] Consider adding transaction management for multi-step operations
- [ ] Replace generic exceptions with specific ones for better error handling
- [ ] Add comprehensive input validation

## Security Concerns
- [x] Missing security flags on cookies (secure, SameSite)
- [x] Consider adding rate limiting for authentication attempts

## Code Quality
- [x] Some methods could benefit from better null checking
- [ ] Consider extracting magic numbers to constants
- [ ] Add more comprehensive JavaDoc comments

## Suggestions
- [x] Consider implementing a strategy pattern for different auth methods
- [ ] Add unit tests for edge cases
- [ ] Consider using builder pattern for complex object creation

Overall, the authentication system looks well-structured. The main areas to focus on are security hardening and transaction management. Let me know if you'd like me to elaborate on any of these points!

File: src/main/java/com/prolinkli/core/app/components/user/service/UserAuthService.java (hunk 4 above)
Comment (Identityofsine) [reply src/main/java/com/prolinkli/core/app/components/user/service/UserAuthService.java:74]: this could be a good todo to start writing validator classes that use a strategy pattern to start checking things; good for sprint 2

File: src/main/java/com/prolinkli/core/app/components/user/service/UserAuthService.java (hunk 5 above)
Comment (Identityofsine) [reply src/main/java/com/prolinkli/core/app/components/user/service/UserAuthService.java:46]: fixed

File: src/main/java/com/prolinkli/framework/auth/util/OAuthUsernameUtil.java [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    // Truncate to a maximum length of the email necessary
+    while ((emailBuilder.length() + oauthIdBuilder.length()) > 30) {
```
[the remark this answers is not in the document]
Comment (Identityofsine) [reply src/main/java/com/prolinkli/framework/auth/util/OAuthUsernameUtil.java:76]: fixed

File: src/main/java/com/prolinkli/framework/auth/util/OAuthUsernameUtil.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+    // Normalize email to lowercase and remove any special characters
+    String normalizedEmail = email.split("@")[0]; // Use the part before '@' for username
+    normalizedEmail = normalizedEmail.toLowerCase().replaceAll("[^a-z0-9]", "");
+
+    String truncatedOauthId = oauthId.replaceAll("[^a-z0-9]", "").toLowerCase().substring(0,
+        Math.min(oauthId.length(), 10));
+
+    StringBuilder emailBuilder = new StringBuilder(normalizedEmail);
+    StringBuilder oauthIdBuilder = new StringBuilder(truncatedOauthId);
```
[the remark this answers is not in the document]
Comment (Identityofsine) [reply src/main/java/com/prolinkli/framework/auth/util/OAuthUsernameUtil.java:72]: will not fix.

File: src/main/java/com/prolinkli/framework/auth/util/OAuthUsernameUtil.java (hunk 15 above)
Comment (Identityofsine) [reply src/main/java/com/prolinkli/framework/auth/util/OAuthUsernameUtil.java]: went to boolean instead.

File: src/main/java/com/prolinkli/framework/cookies/util/JwtCookieUtil.java
[the remark this answers is not in the document]
Comment (Identityofsine) [reply src/main/java/com/prolinkli/framework/cookies/util/JwtCookieUtil.java:1]: all cookie nonseense removed
